src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs

```diff
@@ -15,129 +16,117 @@ namespace Microsoft.CodeAnalysis
+                return true;
             }
 
-            return version != default;
+            version = default;
+            return false;
         }
 
         public override TextAndVersion GetValue(CancellationToken cancellationToken = default)
         {
-            if (_text == null)
+            if (_initialSourceOrRecoverableText is ValueSource<TextAndVersion> source)
             {
-                using (Gate.DisposableWait(cancellationToken))
-                {
-                    if (_text == null)
-                    {
-                        return InitRecoverable(_initialSource!.GetValue(cancellationToken));
-                    }
-                }
+                // replace initial source with recovarable text, if it hasn't been replaced already:
+                Interlocked.CompareExchange(
+                    ref _initialSourceOrRecoverableText,
+                    value: new RecoverableText(source.GetValue(cancellationToken), _services),
+                    comparand: source);
             }
 
-            return TextAndVersion.Create(_text.GetValue(cancellationToken), _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
+            return recoverableText.ToTextAndVersion(recoverableText.GetValue(cancellationToken));
         }
 
         public override async Task<TextAndVersion> GetValueAsync(CancellationToken cancellationToken = default)
         {
-            if (_text == null)
+            if (_initialSourceOrRecoverableText is ValueSource<TextAndVersion> source)
             {
-                using (await Gate.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
-                {
-                    if (_text == null)
-                    {
-                        return InitRecoverable(await _initialSource!.GetValueAsync(cancellationToken).ConfigureAwait(false));
-                    }
-                }
+                // replace initial source with recovarable text, if it hasn't been replaced already:
+                Interlocked.CompareExchange(
+                    ref _initialSourceOrRecoverableText,
+                    value: new RecoverableText(await source.GetValueAsync(cancellationToken).ConfigureAwait(false), _services),
+                    comparand: source);
             }
 
-            var text = await _text.GetValueAsync(cancellationToken).ConfigureAwait(false);
-            return TextAndVersion.Create(text, _version, _filePath, _loadDiagnostic);
+            var recoverableText = (RecoverableText)_initialSourceOrRecoverableText;
+            return recoverableText.ToTextAndVersion(await recoverableText.GetValueAsync(cancellationToken).ConfigureAwait(false));
         }
 
-        private TextAndVersion InitRecoverable(TextAndVersion textAndVersion)
+        private sealed class RecoverableText : WeaklyCachedRecoverableValueSource<SourceText>, ITextVersionable
```

Comment:
WeaklyCachedRecoverableValueSource :'(

i will be happy when we can get rid of all these value-sources.